src/async_traits.rs

```diff
@@ -371,53 +390,86 @@ where
         let async_conn = Connection(Self::as_async_conn(&conn).0.clone());
         match f(async_conn).await {
             Ok(value) => {
-                conn.run_with_shared_connection(|conn| {
-                    Conn::TransactionManager::commit_transaction(conn).map_err(E::from)
-                })
-                .await?;
-                Ok(value)
+                match conn
+                    .run_with_shared_connection(|conn| {
+                        Conn::TransactionManager::commit_transaction(conn)
+                            .map_err(|err| E::from(RunError::User(err)))
+                    })
+                    .await
+                {
+                    Ok(()) => Ok(value),
+                    // XXX: we should try to roll this back
+                    Err(RunError::User(err)) => Err(err),
+                    Err(RunError::RuntimeShutdown) => Err(RunError::RuntimeShutdown.into()),
+                }
             }
             Err(user_error) => {
                 match conn
                     .run_with_shared_connection(|conn| {
-                        Conn::TransactionManager::rollback_transaction(conn).map_err(E::from)
+                        Conn::TransactionManager::rollback_transaction(conn)
+                            .map_err(|err| E::from(RunError::User(err)))
                     })
                     .await
                 {
                     Ok(()) => Err(user_error),
-                    Err(err) => Err(err),
+                    Err(RunError::User(err)) => Err(err),
+                    Err(RunError::RuntimeShutdown) => Err(RunError::RuntimeShutdown.into()),
                 }
             }
         }
     }
 }
 
+fn handle_spawn_blocking_error<T, E>(
```

Comment:
If we're going to be propagating errors out during runtime shutdown, this seems reasonable to me (modulo the comment on whether we can collapse one layer of generics, see my other comment).

However, I'm still a little confused here - before this PR, we had a call-stack of:
- User code
- ... calls async_bb8_diesel API
- ... which tries to spawn a task and fails, leading to a panic

Now, we'll have:
- User code
- ... calls async_bb8_diesel API
- ... which tries to spawn a task and fails, returning an error

But what do we expect the caller to do? In the shutdown case, won't the caller most likely panic anyway?

Put another way -- what useful thing do we expect the caller to do **other** than panic here?

Reply:
Also, to be clear, I'm asking this question not because I believe propagating the error is wrong, but because I fear I may have a gap in my understanding of "why this will make the https://github.com/oxidecomputer/omicron/issues/6505 issues less bad".

Reply:
Ah yeah that's what I was talking about in my comments — the caller would have to learn to squelch the error.

Basically there is some point in the call stack at which the return value would be (), or some other type with a sensible default value. At that point, the caller would have to examine the error and immediately return the default value.

One alternative would be to move the runtime shutdown situation into the Ok case but that feels even weirder.

Definitely not a great situation.

Reply:
I think a useful next step would be to try patching this PR into Omicron and seeing how to manage this situation.